### ISSUE TITLE
Latent Replay for Real-Time Continual Learning

### ISSUE THREAD

**Issue by juanmc2005** [juanmc2005/phd-bib#35]
## [Latent Replay for Real-Time Continual Learning](https://arxiv.org/abs/1912.01100)
*Training deep neural networks at the edge on light computational devices, embedded systems and robotic platforms is nowadays very challenging. Continual learning techniques, where complex models are incrementally trained on small batches of new data, can make the learning problem tractable even for CPU-only embedded devices enabling remarkable levels of adaptiveness and autonomy. However, a number of practical problems need to be solved: catastrophic forgetting before anything else. In this paper we introduce an original technique named "Latent Replay" where, instead of storing a portion of past data in the input space, we store activations volumes at some intermediate layer. This can significantly reduce the computation and storage required by native rehearsal. To keep the representation stable and the stored activations valid we propose to slow-down learning at all the layers below the latent replay one, leaving the layers above free to learn at full pace. In our experiments we show that Latent Replay, combined with existing continual learning techniques, achieves state-of-the-art performance on complex video benchmarks such as CORe50 NICv2 (with nearly 400 small and highly non-i.i.d. batches) and OpenLORIS. Finally, we demonstrate the feasibility of nearly real-time continual learning on the edge through the deployment of the proposed technique on a smartphone device.*

### Cite this paper
```bibtex
@article{pellegrini2019latent,
    title={Latent Replay for Real-Time Continual Learning},
    author={Lorenzo Pellegrini and Gabriele Graffieti and Vincenzo Lomonaco and Davide Maltoni},
    year={2019},
    eprint={1912.01100},
    archivePrefix={arXiv},
    primaryClass={cs.LG}
}
```

**Comment by juanmc2005** [comment]
This paper introduces a novel technique to deal with catastrophic forgetting called _Latent Replay_, which consists in storing activations from a specific layer of the network and perform replay with them, saving computational costs (replay does not happen for only a subset of layers) and storage (higher level activations are smaller in size than inputs).

A problem with this technique is that replay activations and actual activations at the replay layer may differ with time as learning progresses (what they call _aging_). To solve this, they propose to reduce the learning rate of the frontal network (up to the latent replay layer), while leaving the back network with an unrestricted learning rate.

During training, real activations are concatenated to replay activations. Backpropagation is done with all activations for the back network, and with only real activations for the frontal network, as shown in the figure:

![image](https://user-images.githubusercontent.com/5760780/76863880-807c4b80-6860-11ea-814d-64fd283005d8.png)

Using MobileNetV1 with state-of-the-art CL algorithms (CWR*, AR1*, etc) as base architectures and for the CORe50 NICv2 task (new instances and classes), they show that selection of the latent replay layer implies a trade-off between computation, storage and accuracy:
- Computation is reduced when the latent replay layer is closer to the back end of the network
- Storage is reduced when the latent replay layer has few neurons, as these will be stored and replayed
- Accuracy is augmented when the replay layer is closer to the input layer, as this is equivalent to input replay

Moreover, they show that latent activations can be heavily compressed due to sparcity already induced by ReLU, and by using regularization techniques to attract activations of the latent replay layer towards zero, as shown in the literature. They cite work claiming that MobileNet's activations can be compressed by up to 10 times with proper sparsification.

They put their method to the test by training a CL model on an Android app without GPU. They achieve very good results both in terms of performance and resource consumption. However, their work only focuses on new instances and classes, so there's no guarantee that such an approach would perform well on other harder CL benchmarks (like domain change), even though it's promising.

It would be an interesting thing for us to test such a method, given that our goal is to learn representations in a continual fashion. This could provide a proxy for us to evaluate embedding quality, and to decide which embeddings provide the most information while keeping a considerable trade-off on resource consumption.
